File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        int[] buy = new int[k+1];
        int[] sell = new int[k+1];

        Arrays.fill(buy,prices[0]);

        for (int i = 1; i< prices.length; i++){
            for(int j =1; j<= k; j++){
                buy[j] = Math.min(buy[j], prices[i]-sell[j-1]);
                sell[j] = Math.max(sell[j], prices[i]-buy[j]);
            }
            
        }
        return sell[k];
    }
}

File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n = len(prices)
        buy = [prices[0]] * (k+1)
        sell = [0] * (k+1)

        for i in range(1,n):
            for j in range(1,k+1):
                buy[j] = min(buy[j], prices[i] - sell[j-1])
                sell[j] = max(sell[j], prices[i] - buy[j])

        return sell[k]

        